src/Roassal3-Layouts/RSForceLayoutInSpaces.class.st

```diff
@@ -0,0 +1,196 @@
+				| vector u d unit middle left right |
+				vector := a - b.
+				u := vector normal.
+				d := a distanceTo: b.
+				unit := vector / (vector r max: 1).
+				middle := b + (unit * d / 2).
+				left := middle - (u * d * 0.15).
+				right := middle + (u * d * 0.15).
+				result add: ({ left. right } detectMax: #y) afterIndex: 1 ].
+			result ];
+		yourself.
+		
+	RSEdgeBuilder bezier
+		controlPointsController: cpController;
+		canvas: self canvas;
+		color: colorForLinesInSameSpace;
+		shapes: aCollection;
+		connectFrom: #superclass to: #yourself.
+	self canvas addAll: aCollection.
+	layout := (space propertyAt: #layout)
+		nodes: aCollection;
+		edges: self canvas edges;
+		start.
+	aCollection do: [ :s | s propertyAt: #layout put: layout ].
+	aCollection @ (RSDraggableForce new layout: layout).
+	aCollection do:[ :s | s when: RSMouseDragEnd send: #dragEnd: to: self ]
+
+]
+
+{ #category : #initialization }
+RSForceLayoutInSpaces >> initialize [
+	super initialize.
+	colorForLinesInSameSpace := Color blue muchDarker alpha: 0.8.
+	colorForLinesInDifferentSpace := Color blue muchDarker alpha: 0.1.
+	colorScale := NSScale category20.
+	linearScale := NSScale linear domain: #(5 20); range: #(5 10).
+]
+
+{ #category : #shapes }
+RSForceLayoutInSpaces >> newShapeFor: aClass [
+	^ RSEllipse new
+		popup;
+		model: aClass;
+		color: (colorScale scale: aClass package name);
+		size: (linearScale scale: aClass asString size);
+		yourself
+]
+
+{ #category : #shapes }
+RSForceLayoutInSpaces >> newSpaceFor: model [
+	| space layout |
+	layout := RSRectangleForceLayout new
+		start;
+		yourself.
+	space := RSBox new
+		noPaint;
+		model: model;
+		extent: 400@400;
+		border: (RSBorder new color: (colorScale scale: model); width: 2; dashArray: #(10); yourself);
+		propertyAt: #layout put: layout;
+		yourself.
+	self canvas newAnimation repeat onStepDo: [ :t | layout step. ].
+	^ space
+]
+
+{ #category : #accessing }
+RSForceLayoutInSpaces >> numberOfSpaces [
+	^ numberOfSpaces
+]
+
+{ #category : #accessing }
+RSForceLayoutInSpaces >> numberOfSpaces: anInteger [ 
+	numberOfSpaces := anInteger
+]
+
+{ #category : #hooks }
+RSForceLayoutInSpaces >> renderIn: aCanvas [
+	spaces := (1 to: numberOfSpaces) collect: [ :i | self newSpaceFor: i ].
+	shapes := classes collect: [ :obj | self newShapeFor: obj ] as: RSGroup.
+	self fill: spaces first with: shapes.
+	aCanvas addAll: spaces.
+	(RSGridLayout withGap: 20) on: spaces.
+	spaces do: [ :space | self updateSpace: space. ].
+	aCanvas @ RSSimpleCanvasMonitor.
+	aCanvas @ RSCanvasController.
+]
+
+{ #category : #hooks }
+RSForceLayoutInSpaces >> updateSpace: space [
+	| layout |
+	layout := space propertyAt: #layout.
+	layout center: space position.
+	layout enclosingRectangle: space encompassingRectangle
+]
+
+{ #category : #hooks }
+RSForceLayoutInSpaces >> withShapesThatAccompany: shape [
+	^ shape model withAllSubclasses collect: [ :cls | self canvas shapeFromModel: cls ].
```

Comment:
We should either rename this class to *Example or create that class by separate, and extract there come like this which is very example/model specific.